### PR TITLE
Improve repo configuration based on scientific python suggestions 

### PR DESCRIPTION
**Pull request by Czaki**
# Description

The Scientific python has tool named repo-review with set of rules that they think are good for projects.

Not all apply to napari, but some of them looks usefull. To see the check on napari repo one may use this link 
https://learn.scientific-python.org/development/guides/repo-review/?repo=napari%2Fnapari&branch=main

This PR contains non mypy improvements, as mypy ones requires improvements in code. 

It depends on #6557

This PR also enables automatically dump graph when QtViewer is leaked during tests. Then such graph in pdf is uploaded as job artifact.

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -103,6 +103,7 @@ commands =
     !cov: python \
         -m pytest {env:PYTEST_PATH:} --color=yes --basetemp={envtmpdir} \
         --ignore tools --maxfail=5 --json-report \
+    linux: --pystack-threshold=60 --pystack-args="--native-all" \
```
Comment (brisvag) [inline tox.ini:108]: Why this change? I guess it's because of the added `-ra` option maybe?

**Review by brisvag — APPROVED** [review]

File: pyproject.toml [hunk 2]
```diff
@@ -151,10 +148,12 @@ xarray = "xr"
 [tool.pytest.ini_options]
 # These follow standard library warnings filters syntax.  See more here:
 # https://docs.python.org/3/library/warnings.html#describing-warning-filters
-addopts = "--maxfail=5 --durations=10 -rXxs"
+addopts = ["--maxfail=5",  "--durations=10",  "-ra", "--strict-markers", "--strict-config"]
```
Comment (Czaki) [inline pyproject.toml:345]: @brisvag Asked changes comes from using `--strict-config` here. 

It causes pytest to fail if some option is not recognized. It makes harder to configure optional plugins, but save from making typos in configuration options.

Here issue in pytest-pystack https://github.com/bloomberg/pytest-pystack/issues/9

File: napari/layers/labels/_tests/test_labels.py [hunk 3]
```diff
@@ -1123,9 +1123,6 @@ def test_cursor_size_with_negative_scale():
     assert layer.cursor_size > 0
 
 
-@pytest.mark.xfail(
```
Comment (psobolewskiPhD) [inline napari/layers/labels/_tests/test_labels.py:1079]: what's motivating this change?

**Comment by psobolewskiPhD** [comment]
Ran the repo-review tool locally on this PR branch.
Looks good. 
```
General:

 • Detected build backend: setuptools.build_meta                                

├── PY001 Has a pyproject.toml ✅
├── PY002 Has a README.(md|rst) file ✅
├── PY003 Has a LICENSE* file ✅
├── PY004 Has docs folder ✅
├── PY005 Has tests folder ✅
├── PY006 Has pre-commit config ✅
└── PY007 Supports an easy task runner (nox or tox) ✅

PyProject:
├── PP002 Has a proper build-system table ✅
├── PP003 Does not list wheel as a build-dep ✅
├── PP301 Has pytest in pyproject ✅
├── PP302 Sets a minimum pytest to at least 6 ✅
├── PP303 Sets the test paths ✅
├── PP304 Sets the log level in pytest ✅
├── PP305 Specifies xfail_strict ✅
├── PP306 Specifies strict config ✅
├── PP307 Specifies strict markers ✅
├── PP308 Specifies useful pytest summary ✅
└── PP309 Filter warnings specified ✅

GitHub Actions:
├── GH100 Has GitHub Actions config ✅
├── GH101 Has nice names ✅
├── GH102 Auto-cancel on repeated PRs ✅
├── GH103 At least one workflow with manual dispatch trigger ✅
├── GH104 Use unique names for upload-artifact ✅
├── GH200 Maintained by Dependabot ✅
├── GH210 Maintains the GitHub action versions with Dependabot ✅
├── GH211 Do not pin core actions as major versions ✅
└── GH212 Require GHA update grouping ✅

```

File: napari/layers/labels/_tests/test_labels.py (hunk 3 above)
Comment (Czaki) [reply napari/layers/labels/_tests/test_labels.py:1079]: using `xfail_strict = true` rule in pytest configuration (PP305)

**Review by brisvag — DISMISSED** [review]
Nice work! Could you add some details to the description about non-obvious changes, like the "leaked viewer graph" part and testsupport changes?

**Comment by Czaki** [comment]
I just realized that min req is still failing.

File: .github/workflows/test_pull_requests.yml [hunk 4]
```diff
@@ -88,15 +88,23 @@ jobs:
...
+            backend: pyqt5
+            MIN_REQ: 1
+            coverage: cov
     with:
       python_version: ${{ matrix.python }}
       platform: ${{ matrix.platform }}
       qt_backend: ${{ matrix.backend }}
-      coverage: no_cov
+      coverage: ${{ matrix.coverage }}
+      min_req: ${{ matrix.MIN_REQ }}
```
Comment (Czaki) [inline .github/workflows/test_pull_requests.yml:102]: This is to simplify copy and paste problematic job to initial section and speedup CI.

File: napari/utils/_testsupport.py [hunk 5]
```diff
@@ -56,16 +57,22 @@ def fail_obj_graph(Klass):
...
             list(Klass._instances),
-            max_depth=20,
-            filename=f'{Klass.__name__}-leak-backref-graph-{COUNTER}.pdf',
+            max_depth=40,
+            filename=str(file_path),
         )
 
+        Klass._instances.clear()
+
+        assert file_path.exists()
```
Comment (Czaki) [inline napari/utils/_testsupport.py]: ensure proper path for save of leaked graph

File: .github/workflows/reusable_run_tox_test.yml [hunk 6]
```diff
@@ -68,6 +70,9 @@ jobs:
 
       - uses: tlambert03/setup-qt-libs@v1
 
+      - name: Setup Graphviz
+        uses: ts-graphviz/setup-graphviz@v2
```
Comment (Czaki) [inline .github/workflows/reusable_run_tox_test.yml:74]: Required to generate graph of leaked objects

File: .github/workflows/reusable_run_tox_test.yml [hunk 7]
```diff
@@ -178,6 +183,19 @@ jobs:
...
+        run: |
+          pwd
+          ls -la
+
+      - name: Upload leaked viewer graph
+        if: failure()
+        uses: actions/upload-artifact@v4
+        with:
+          name: leaked ${{ inputs.platform }} py ${{ inputs.python_version }} ${{ inputs.toxenv || inputs.qt_backend }}
+          path: ./*leak-backref-graph*.pdf
```
Comment (Czaki) [inline .github/workflows/reusable_run_tox_test.yml:193]: it upload graph of leaked objects if such happen in tests

File: tox.ini [hunk 8]
```diff
@@ -84,6 +84,7 @@ deps =
     pyside6: PySide6 != 6.4.3, !=6.5.0 ; python_version >= '3.10'
     pytest-json-report
     pytest-pystack ; sys_platform == 'linux'
+    objgraph
```
Comment (Czaki) [inline tox.ini:86]: Required to dump information about leaked objects

File: tox.ini [hunk 9]
```diff
@@ -93,6 +94,7 @@ extras =
 allowlist_externals =
     echo
     mypy
+    dot
```
Comment (Czaki) [inline tox.ini:97]: Required for leaked object graph generation

**Review by Czaki — COMMENTED** [review]
@brisvag You mean such comments or in code?

**Comment by brisvag** [comment]
I guess the only new thing is the graph of leaks then. Yeah just add a note in the PR description that this was added :)

**Comment by brisvag** [comment]
I'll mark my review as stale since you're still doing lots of testing and I wanna avoid accidental merge. Ping me when you think things are ok!

File: pyproject.toml [hunk 10]
```diff
@@ -344,10 +342,12 @@ xarray = "xr"
...
 console_output_style = "count"
-pystack_threshold=60
-pystack_args="--native-all"
+minversion = "7"
```
Comment (Czaki) [inline pyproject.toml]:
```suggestion
minversion = "8"
```

File: pyproject.toml [hunk 11]
```diff
@@ -344,10 +342,12 @@ xarray = "xr"
 [tool.pytest.ini_options]
 # These follow standard library warnings filters syntax.  See more here:
 # https://docs.python.org/3/library/warnings.html#describing-warning-filters
-addopts = "--maxfail=5 --durations=10 -rXxs"
+addopts = ["--maxfail=5",  "--durations=10",  "-ra", "--strict-markers", "--strict-config"]
 console_output_style = "count"
-pystack_threshold=60
-pystack_args="--native-all"
+minversion = "8"
+log_cli_level = "INFO"
```
Comment (Czaki) [inline pyproject.toml]:
```suggestion
# log_cli_level = "INFO"
```

**Comment by Czaki** [comment]
@brisvag I found that the problem was caused by logging level (I do not know why). But now it is ready for review, and debug of logging will be another PR.

**Review by brisvag — APPROVED** [review]
